File: python/elizaos_plugin_polymarket/actions/realtime.py
"""
Real-time WebSocket actions for Polymarket.
"""

import os
from typing import Protocol

from elizaos_plugin_polymarket.error import PolymarketError, PolymarketErrorCode


class RuntimeProtocol(Protocol):
    """Protocol for agent runtime."""

    def get_setting(self, key: str) -> str | None:
        """Get a setting value."""
        ...


async def setup_websocket(
    channels: list[str] | None = None,
    asset_ids: list[str] | None = None,
    authenticated: bool = False,
    runtime: RuntimeProtocol | None = None,
) -> dict[str, object]:
    """
    Setup and configure WebSocket connections for real-time data.

    Args:
        channels: List of channels to subscribe to (e.g., ['book', 'price', 'trade'])
        asset_ids: List of asset IDs to subscribe to
        authenticated: Whether to use authenticated WebSocket connection
        runtime: Optional agent runtime for settings

    Returns:
        Dictionary with WebSocket configuration

    Raises:
        PolymarketError: If WebSocket setup fails
    """
    try:
        clob_ws_url = (
            runtime.get_setting("CLOB_WS_URL") if runtime else os.environ.get("CLOB_WS_URL")
        ) or (runtime.get_setting("CLOB_API_URL") if runtime else os.environ.get("CLOB_API_URL"))

        if not clob_ws_url:
            raise PolymarketError(
                PolymarketErrorCode.CONFIG_ERROR,
                "CLOB_WS_URL or CLOB_API_URL is required for WebSocket connections",
            )

        # Convert HTTP URL to WebSocket URL if needed
        if clob_ws_url.startswith("http://"):
            ws_url = clob_ws_url.replace("http://", "ws://")
        elif clob_ws_url.startswith("https://"):
            ws_url = clob_ws_url.replace("https://", "wss://")
        elif not clob_ws_url.startswith(("ws://", "wss://")):
            ws_url = f"wss://{clob_ws_url}"
        else:
            ws_url = clob_ws_url

        default_channels = channels or ["book", "price"]
        asset_ids_list = asset_ids or []

        # Check if authenticated credentials are available
        has_credentials = False
        if authenticated:
            clob_api_key = (
                runtime.get_setting("CLOB_API_KEY") if runtime else os.environ.get("CLOB_API_KEY")
            )
            clob_api_secret = (
                runtime.get_setting("CLOB_API_SECRET")
                if runtime
                else os.environ.get("CLOB_API_SECRET")
            ) or (runtime.get_setting("CLOB_SECRET") if runtime else os.environ.get("CLOB_SECRET"))
            clob_api_passphrase = (
                runtime.get_setting("CLOB_API_PASSPHRASE")
                if runtime
                else os.environ.get("CLOB_API_PASSPHRASE")
            ) or (
                runtime.get_setting("CLOB_PASS_PHRASE")
                if runtime
                else os.environ.get("CLOB_PASS_PHRASE")
            )
            has_credentials = bool(clob_api_key and clob_api_secret and clob_api_passphrase)

        return {
            "url": ws_url,
            "channels": default_channels,
            "asset_ids": asset_ids_list,
            "authenticated": authenticated and has_credentials,
            "status": "disconnected",  # Would be 'connected' if service is running
            "has_credentials": has_credentials,
        }

    except PolymarketError:
        raise
    except Exception as e:
        raise PolymarketError(
            PolymarketErrorCode.WEBSOCKET_ERROR,
            f"Failed to setup WebSocket: {e}",
            cause=e,
        ) from e


async def handle_realtime_updates(
    action: str = "status",
    channel: str | None = None,
    asset_ids: list[str] | None = None,
    runtime: RuntimeProtocol | None = None,
) -> dict[str, object]:
    """
    Handle real-time WebSocket updates (subscribe/unsubscribe/status).

    Args:
        action: Action to perform ('subscribe', 'unsubscribe', or 'status')
        channel: Channel name (e.g., 'price', 'book', 'trade')
        asset_ids: List of asset IDs for subscription
        runtime: Optional agent runtime for settings

    Returns:
        Dictionary with subscription status

    Raises:
        PolymarketError: If handling updates fails
    """
    try:
        clob_ws_url = (
            runtime.get_setting("CLOB_WS_URL") if runtime else os.environ.get("CLOB_WS_URL")
        ) or (runtime.get_setting("CLOB_API_URL") if runtime else os.environ.get("CLOB_API_URL"))

        if not clob_ws_url:
            raise PolymarketError(
                PolymarketErrorCode.CONFIG_ERROR,
                "CLOB_WS_URL or CLOB_API_URL is required for WebSocket connections",
            )

        # Available channels
        available_channels = {
            "price": "Real-time price updates",
            "book": "Order book changes",
            "trade": "Trade executions",
            "ticker": "Market ticker updates",
            "user": "Authenticated user updates (orders, fills)",
        }

        if action == "status":
            return {
                "ws_url": clob_ws_url,
                "available_channels": available_channels,
                "subscriptions": [],  # Would contain active subscriptions if service is running
                "status": "disconnected",  # Would be 'connected' if service is running
            }
        elif action == "subscribe":
            if not channel:
                raise PolymarketError(
                    PolymarketErrorCode.WEBSOCKET_ERROR,
                    "Channel is required for subscription",
                )
            if channel not in available_channels:
                raise PolymarketError(
                    PolymarketErrorCode.WEBSOCKET_ERROR,
                    f"Invalid channel: {channel}. Available: {', '.join(available_channels.keys())}",
                )
            if not asset_ids:
                raise PolymarketError(
                    PolymarketErrorCode.WEBSOCKET_ERROR,
                    "At least one asset ID is required for subscription",
                )

            return {
                "action": "subscribe",
                "channel": channel,
                "asset_ids": asset_ids,
                "status": "pending",  # Would be 'active' if service is running
            }
        elif action == "unsubscribe":
            if not channel:
                raise PolymarketError(
                    PolymarketErrorCode.WEBSOCKET_ERROR,
                    "Channel is required for unsubscription",
                )

            return {
                "action": "unsubscribe",
                "channel": channel,
                "asset_ids": asset_ids,
                "status": "pending",  # Would be 'unsubscribed' if service is running
            }
        else:
            raise PolymarketError(
                PolymarketErrorCode.WEBSOCKET_ERROR,
                f"Invalid action: {action}. Must be 'subscribe', 'unsubscribe', or 'status'",
            )

    except PolymarketError:
        raise
    except Exception as e:
        raise PolymarketError(
            PolymarketErrorCode.WEBSOCKET_ERROR,
            f"Failed to handle realtime updates: {e}",
            cause=e,
        ) from e

File: rust/src/actions/realtime.rs
#![allow(missing_docs)]
//! Real-time WebSocket actions for Polymarket

use crate::client::ClobClient;
use crate::error::{PolymarketError, Result};

/// WebSocket subscription status
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SubscriptionStatus {
    /// Disconnected
    Disconnected,
    /// Connecting
    Connecting,
    /// Connected
    Connected,
    /// Error state
    Error,
}

/// WebSocket configuration
#[derive(Debug, Clone)]
pub struct WebsocketConfig {
    /// WebSocket URL
    pub url: String,
    /// Channels to subscribe to
    pub channels: Vec<String>,
    /// Asset IDs to subscribe to
    pub asset_ids: Vec<String>,
    /// Whether authentication is required
    pub authenticated: bool,
    /// Connection status
    pub status: SubscriptionStatus,
}

/// Setup WebSocket connection for real-time updates
///
/// # Arguments
///
/// * `client` - The CLOB client
/// * `ws_url` - WebSocket URL
/// * `channels` - Channels to subscribe to (e.g., "book", "price", "trade")
/// * `asset_ids` - Asset IDs to subscribe to
/// * `authenticated` - Whether to use authenticated connection
///
/// # Errors
///
/// Returns an error if configuration is invalid
pub fn setup_websocket(
    _client: &ClobClient,
    ws_url: &str,
    channels: &[String],
    asset_ids: &[String],
    authenticated: bool,
) -> Result<WebsocketConfig> {
    if ws_url.is_empty() {
        return Err(PolymarketError::config_error("WebSocket URL is required"));
    }

    if channels.is_empty() {
        return Err(PolymarketError::config_error(
            "At least one channel must be specified",
        ));
    }

    // Validate channel names
    let valid_channels = ["book", "price", "trade", "ticker", "user"];
    for channel in channels {
        if !valid_channels.contains(&channel.as_str()) {
            return Err(PolymarketError::config_error(format!(
                "Invalid channel: {channel}. Valid channels are: {}",
                valid_channels.join(", ")
            )));
        }
    }

    Ok(WebsocketConfig {
        url: ws_url.to_string(),
        channels: channels.to_vec(),
        asset_ids: asset_ids.to_vec(),
        authenticated,
        status: SubscriptionStatus::Disconnected,
    })
}

/// Handle real-time updates subscription management
///
/// # Arguments
///
/// * `client` - The CLOB client
/// * `action` - Action to perform ("subscribe", "unsubscribe", or "status")
/// * `channel` - Channel name (required for subscribe/unsubscribe)
/// * `asset_ids` - Asset IDs (required for subscribe)
///
/// # Errors
///
/// Returns an error if the operation fails
pub fn handle_realtime_updates(
    _client: &ClobClient,
    action: &str,
    channel: Option<&str>,
    asset_ids: Option<&[String]>,
) -> Result<String> {
    match action {
        "subscribe" => {
            let channel = channel.ok_or_else(|| {
                PolymarketError::config_error("Channel is required for subscription")
            })?;
            let asset_ids = asset_ids.ok_or_else(|| {
                PolymarketError::config_error("Asset IDs are required for subscription")
            })?;

            if asset_ids.is_empty() {
                return Err(PolymarketError::config_error(
                    "At least one asset ID is required for subscription",
                ));
            }

            Ok(format!(
                "Subscribed to {} channel for assets: {}",
                channel,
                asset_ids.join(", ")
            ))
        }
        "unsubscribe" => {
            let channel = channel.ok_or_else(|| {
                PolymarketError::config_error("Channel is required for unsubscription")
            })?;
            Ok(format!("Unsubscribed from {} channel", channel))
        }
        "status" => Ok(
            "WebSocket status: Use PolymarketService for actual connection management".to_string(),
        ),
        _ => Err(PolymarketError::config_error(format!(
            "Invalid action: {action}. Valid actions are: subscribe, unsubscribe, status"
        ))),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::client::ClobClient;

    fn test_client() -> ClobClient {
        futures::executor::block_on(ClobClient::new(None, &format!("0x{}", "11".repeat(32))))
            .expect("client init")
    }

    #[test]
    fn test_setup_websocket_empty_url() {
        let client = test_client();
        let err = setup_websocket(&client, "", &[], &[], false).unwrap_err();
        assert!(err.to_string().contains("WebSocket URL is required"));
    }

    #[test]
    fn test_setup_websocket_invalid_channel() {
        let client = test_client();

        let channels = vec!["nope".to_string()];
        let err = setup_websocket(&client, "wss://example.com/ws", &channels, &[], false).unwrap_err();
        assert!(err.to_string().contains("Invalid channel"));
    }

    #[test]
    fn test_handle_realtime_updates_invalid_action() {
        let client = test_client();

        let err = handle_realtime_updates(&client, "nope", None, None).unwrap_err();
        assert!(err.to_string().contains("Invalid action"));
    }
}
